### PR TITLE
feat(new-preference): Add preference toggle for gas denom to be used when making transactions

### PR DESCRIPTION
**Pull request by jaredvu**
https://github.com/dydxprotocol/v4-web/assets/13111994/b0b4fe4d-e28e-4c5b-a5f6-4143cc99af9c

Expose UI for setting a `selectedGasDenom` within an instance of the compositeClient.

---

## Views

* New: `<ViewComponentName>`
  * 
  * 

* `<PreferenceDialog>`
  * Add Switch toggle for gasDenom

## Constants/Types

* `constants/localStorage`
  * add localStorage key for `selectedGasDenom`

## Hooks

* `hooks/useDydxClient`
  * add local storage state and setState for toggling `selectedGasDenom`

* `hooks/useSubaccount`
  * Add native token faucet call 

## State

* `state/configs`
  * Add new preference key

## Packages

* `v4-client-js`
  * Adds ability to toggle gas denom
  * updated: v1.1.8 -> v1.1.10

---

### CODE REVIEW

File: src/hooks/useSubaccount.tsx [hunk 1]
```diff
@@ -463,17 +467,15 @@ export const useSubaccountContext = ({ localDydxWallet }: { localDydxWallet?: Lo
         parsingError?: Nullable<ParsingError>,
         data?: Nullable<HumanReadableTriggerOrdersPayload>
       ) => {
-        const placeOrderPayloads = data?.placeOrderPayloads.toArray() || []
-        const cancelOrderPayloads = data?.cancelOrderPayloads.toArray() || []
+        const placeOrderPayloads = data?.placeOrderPayloads.toArray() || [];
+        const cancelOrderPayloads = data?.cancelOrderPayloads.toArray() || [];
```
Comment (yogurtandjam) [inline src/hooks/useSubaccount.tsx:471]: 🥳

File: src/hooks/useDydxClient.tsx [hunk 2]
```diff
@@ -273,6 +293,10 @@ const useDydxClientContext = () => {
     [compositeClient]
   );
 
+  const getAllValidators = useCallback(async () => {
```
Comment (yogurtandjam) [inline src/hooks/useDydxClient.tsx:296]: [q] was this intentional? doesn't seem related to the rest of the pr

File: src/hooks/useSubaccount.tsx [hunk 3]
```diff
@@ -66,7 +66,7 @@ export const useSubaccountContext = ({ localDydxWallet }: { localDydxWallet?: Lo
   const { usdcDenom, usdcDecimals } = useTokenConfigs();
   const { compositeClient, faucetClient } = useDydxClient();
 
-  const { getFaucetFunds } = useMemo(
+  const { getFaucetFunds, getNativeTokens } = useMemo(
```
Comment (yogurtandjam) [inline src/hooks/useSubaccount.tsx:69]: [q] is this related to the gas denom toggle functionality? not seeing this used anywhere else

File: src/hooks/useDydxClient.tsx (hunk 2 above)
Comment (aforaleka) [reply src/hooks/useDydxClient.tsx:296]: hahaha are we just exposing this function for candidates x_x

**Review by yogurtandjam — APPROVED** [review]
general code LGTM, just some questions

File: src/hooks/useSubaccount.tsx (hunk 3 above)
Comment (aforaleka) [reply src/hooks/useSubaccount.tsx:69]: this is called in `requestFaucetFunds` / when you "desposit" on testnet via faucet

File: src/hooks/useSubaccount.tsx [hunk 4]
```diff
@@ -346,11 +349,12 @@ export const useSubaccountContext = ({ localDydxWallet }: { localDydxWallet?: Lo
 
     try {
       await getFaucetFunds({ dydxAddress, subaccountNumber });
+      await getNativeTokens({ dydxAddress });
```
Comment (aforaleka) [inline src/hooks/useSubaccount.tsx:352]: can they be done in parallel? can we do something like
`await Promise.all([getFaucetFunds(...), getNativeTokens(...)])`

File: src/hooks/useDydxClient.tsx (hunk 2 above)
Comment (jaredvu) [reply src/hooks/useDydxClient.tsx:296]: WOOOPS! I wanted to see the payload to help answer some questions. will remove 😂

File: src/hooks/useSubaccount.tsx (hunk 4 above)
Comment (jaredvu) [reply src/hooks/useSubaccount.tsx:352]: Ah yeah! Will add to promise.all

File: src/hooks/useDydxClient.tsx [hunk 5]
```diff
@@ -112,6 +115,23 @@ const useDydxClientContext = () => {
...
   }, [networkConfig]);
 
+  // ------ Gas Denom ------ //
+
+  const [gasDenom, setGasDenom] = useLocalStorage<SelectedGasDenom>({
+    key: LocalStorageKey.SelectedGasDenom,
+    defaultValue: SelectedGasDenom.USDC,
+  });
+
+  const setSelectedGasDenom = useCallback(
```
[the remark this answers is not in the document]
Comment (jaredvu) [reply src/hooks/useDydxClient.tsx:125]: I used `useCallback` because this fn is a dependency in a `useMemo` in `<PreferencesDialog>`. I wanted the `useMemo` to update when `compositeClient` initializes but no further re-renders after.

I could either keep `useCallback` here, or add `compositeClient` as the dep in `<PreferencesDialog>`

File: src/hooks/useSubaccount.tsx [hunk 6]
```diff
@@ -534,7 +536,7 @@ export const useSubaccountContext = ({ localDydxWallet }: { localDydxWallet?: Lo
       } else if (!localDydxWallet) {
         throw new Error('wallet not initialized');
       } else if (!newMarketProposal) {
-        throw new Error('governance variables not initialized');
+        throw new Error(' governance variables not initialized');
```
Comment (aforaleka) [inline src/hooks/useSubaccount.tsx:539]: ummmmmmm

**Review by aforaleka — APPROVED** [review]
